### PR TITLE
feat(frontend): optimize constant project with exact input row count known into value operator

### PR DESCRIPTION
**Pull request by jon-chuang**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
optimize constant project into value operator

## Checklist

- [X] I have written necessary rustdoc comments
- [X] I have added necessary unit tests and integration tests
- [X] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).
- [X] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [X] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [X] My PR **DOES NOT** contain user-facing changes.

### Fixes
https://github.com/risingwavelabs/risingwave/issues/7858

### CODE REVIEW

File: src/frontend/planner_test/tests/testdata/index_selection.yaml [hunk 1]
```diff
@@ -85,11 +85,8 @@
     create index idx2 on t1(a) include(a);
     select count(1) from t1;
   batch_plan: |
-    BatchSimpleAgg { aggs: [sum0(count(1:Int32))] }
-    └─BatchExchange { order: [], dist: Single }
-      └─BatchSimpleAgg { aggs: [count(1:Int32)] }
-        └─BatchProject { exprs: [1:Int32] }
-          └─BatchScan { table: idx2, columns: [], distribution: SomeShard }
+    BatchSimpleAgg { aggs: [count(1:Int32)] }
```
Comment (jon-chuang) [inline src/frontend/planner_test/tests/testdata/index_selection.yaml:88]: Is this correct? Seems like it is. The batchsimpleagg is simplified to evaluating a const expr, so it has no input depdendence, so it becomes distribution single. Great!

Comment (st1page) [reply src/frontend/planner_test/tests/testdata/index_selection.yaml:88]: I think it is not correct :hot_face: 
the `count(1)` and `count(*)` do not have input columns but it depend on the input relations, which need to give the row count of the relation.

File: src/frontend/src/optimizer/rule/trivial_project_to_values_rule.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+// limitations under the License.
+
+use super::{BoxedRule, Rule};
+use crate::optimizer::plan_node::LogicalValues;
+use crate::optimizer::PlanRef;
+
+pub struct TrivialProjectToValuesRule {}
+impl Rule for TrivialProjectToValuesRule {
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let project = plan.as_logical_project()?;
```
Comment (st1page) [inline src/frontend/src/optimizer/rule/trivial_project_to_values_rule.rs:22]: maybe we can only do the optimization when the logical project's input is a values operator or other operators which has the determined output row count

File: src/frontend/src/optimizer/rule/trivial_project_to_values_rule.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+                LogicalValues::new(
+                    vec![project.exprs().clone()],
+                    project.schema().clone(),
+                    project.ctx(),
+                )
+                .into(),
+            )
+        } else {
+            None
+        }
```
Comment (chenzl25) [inline src/frontend/src/optimizer/rule/trivial_project_to_values_rule.rs:34]: We need to ensure project's input produces exact one row.

Comment (jon-chuang) [reply src/frontend/src/optimizer/rule/trivial_project_to_values_rule.rs:34]: Hmm, true, else the number of rows can be dynamic...

File: src/frontend/src/optimizer/rule/trivial_project_to_values_rule.rs (hunk 2 above)
Comment (jon-chuang) [reply src/frontend/src/optimizer/rule/trivial_project_to_values_rule.rs:22]: We need to know the exact row count. For instance, for a project on simple agg, will the result of simple agg be `None`?

Comment (jon-chuang) [reply src/frontend/src/optimizer/rule/trivial_project_to_values_rule.rs:22]: Let's just stick to values for now :cry:

Comment (jon-chuang) [reply src/frontend/src/optimizer/rule/trivial_project_to_values_rule.rs:22]: I think I will use a row counter.

File: src/frontend/planner_test/tests/testdata/cast.yaml [hunk 4]
```diff
@@ -3,23 +3,17 @@
   sql: |
     select 1 having NULL;
   batch_plan: |
-    BatchProject { exprs: [1:Int32] }
-    └─BatchFilter { predicate: null:Boolean }
-      └─BatchSimpleAgg { aggs: [] }
-        └─BatchValues { rows: [[]] }
+    BatchValues { rows: [[1:Int32]] }
```
Comment (jon-chuang) [inline src/frontend/planner_test/tests/testdata/cast.yaml:6]: Why is this correct? It should evaluate count of `BatchFilter` to false...

File: src/frontend/planner_test/tests/testdata/index_selection.yaml (hunk 1 above)
Comment (jon-chuang) [reply src/frontend/planner_test/tests/testdata/index_selection.yaml]: fixed

File: src/frontend/src/optimizer/plan_visitor/max_one_row_visitor.rs [hunk 5]
```diff
@@ -92,3 +96,42 @@ impl PlanVisitor<bool> for HasMaxOneRowApply {
         plan.max_one_row() | self.visit(plan.left()) | self.visit(plan.right())
     }
 }
+
+pub struct CountRows;
+
+impl PlanVisitor<Option<usize>> for CountRows {
```
Comment (jon-chuang) [inline src/frontend/src/optimizer/plan_visitor/max_one_row_visitor.rs:102]: Can we also add count for Agg? From what I know, agg has a possibility of produce 0 rows, so no.

Comment (chenzl25) [reply src/frontend/src/optimizer/plan_visitor/max_one_row_visitor.rs:103]: Agg without group key always produces exact one row.

Comment (jon-chuang) [reply src/frontend/src/optimizer/plan_visitor/max_one_row_visitor.rs:103]: Hmm, ok, in that case we can count it as well, let me add test with 0 values below the agg

Comment (jon-chuang) [reply src/frontend/src/optimizer/plan_visitor/max_one_row_visitor.rs:103]: Oops, we have a bug:
```sql
dev=> create table(v1 int);
CREATE_TABLE
dev=> create materialized view v2 as select sum(v1) from t;
CREATE_MATERIALIZED_VIEW
dev=> select * from v2;
 sum 
-----
(0 rows)

dev=> 
```

Comment (jon-chuang) [reply src/frontend/src/optimizer/plan_visitor/max_one_row_visitor.rs:103]: Oops :dizzy_face: 
```sql
dev=> create materialized view v3 as select count(*) from t;
CREATE_MATERIALIZED_VIEW
dev=> select * from v3;
 count 
-------
(0 rows)
```

Comment (jon-chuang) [reply src/frontend/src/optimizer/plan_visitor/max_one_row_visitor.rs:103]: Oops
```sql
dev=> create materialized view v4 as select sum(v1) from t group by v1;
CREATE_MATERIALIZED_VIEW
dev=> select * from v4;
 sum 
-----
(0 rows)
```

Comment (jon-chuang) [reply src/frontend/src/optimizer/plan_visitor/max_one_row_visitor.rs:103]: I will fix it.

Comment (chenzl25) [reply src/frontend/src/optimizer/plan_visitor/max_one_row_visitor.rs:103]: For `SimpleAgg`, We have a known bug which is that it misses the first initial row.

Comment (jon-chuang) [reply src/frontend/src/optimizer/plan_visitor/max_one_row_visitor.rs:103]: Fixed: https://github.com/risingwavelabs/risingwave/pull/7932

Let's merge that and then count the exact rows for simple agg here.

**Review by fuyufjh — APPROVED** [review]
LGTM

**Review by chenzl25 — APPROVED** [review]
LGTM

**Pull request merged**